framework/auth/views.py

```diff
@@ -237,36 +238,43 @@ def login_and_register_handler(auth, login=True, campaign=None, next_url=None, l
                 else:
                     data['next_url'] = cas.get_login_url(next_url, campaign='institution')
             # for non-institution campaigns
-            elif auth.logged_in:
-                # if user is already logged in, go to the campaign landing page
-                data['status_code'] = http.FOUND
-                data['next_url'] = campaigns.campaign_url_for(campaign)
             else:
-                # if user is logged out, go to the osf register page with campaign context
-                if login:
-                    # `GET /login?campaign=...`
-                    data['next_url'] = web_url_for('auth_register', campaign=campaign)
+                destination = next_url if next_url else campaigns.campaign_url_for(campaign)
+                if auth.logged_in:
+                    # if user is already logged in, go to the campaign landing page
+                    data['status_code'] = http.FOUND
+                    data['next_url'] = destination
                 else:
-                    # `GET /register?campaign=...`
-                    data['campaign'] = campaign
-                    if campaigns.is_proxy_login(campaign):
-                        data['next_url'] = web_url_for(
-                            'auth_login',
-                            next=campaigns.campaign_url_for(campaign),
-                            _absolute=True
-                        )
+                    # if user is logged out, go to the osf register page with campaign context
+                    if login:
+                        # `GET /login?campaign=...`
+                        data['next_url'] = web_url_for('auth_register', campaign=campaign, next=destination)
                     else:
-                        data['next_url'] = campaigns.campaign_url_for(campaign)
+                        # `GET /register?campaign=...`
+                        data['campaign'] = campaign
+                        if campaigns.is_proxy_login(campaign):
+                            data['next_url'] = web_url_for(
+                                'auth_login',
+                                next=destination,
+                                _absolute=True
+                            )
+                        else:
+                            data['next_url'] = destination
         else:
             # invalid campaign
             raise HTTPError(http.BAD_REQUEST)
     # login or register with next parameter
     elif next_url:
-        if logout and auth.logged_in:
-            # handle `claim_user_registered`, TODO [#OSF-6998]: talk to product about the `must_login_warning`
-            data['status_code'] = 'auth_logout'
-            data['next_url'] = request.url
-            data['must_login_warning'] = True
+        if logout:
```

Comment:
During writing the unit tests, I found a bug that was introduced by myself. Now fixed and fully tested. Please refer to https://github.com/CenterForOpenScience/osf.io/pull/6545/files#diff-a4f5c4d64d4811b15e1405573c00f3f4R3777.